File: ts-sdk/src/cel/lexer.ts
/**
 * CEL Lexer - tokenize CEL expressions
 */

export type TokenType =
  | 'NUMBER'
  | 'STRING'
  | 'BOOL'
  | 'NULL'
  | 'IDENT'
  | 'DOT'
  | 'COMMA'
  | 'LPAREN'
  | 'RPAREN'
  | 'LBRACKET'
  | 'RBRACKET'
  | 'PLUS'
  | 'MINUS'
  | 'STAR'
  | 'SLASH'
  | 'PERCENT'
  | 'EQ'
  | 'NEQ'
  | 'LT'
  | 'LTE'
  | 'GT'
  | 'GTE'
  | 'AND'
  | 'OR'
  | 'NOT'
  | 'QUESTION'
  | 'COLON'
  | 'IN'
  | 'EOF';

export interface Token {
  type: TokenType;
  value: string | boolean | null;
  pos: number;
}

const KEYWORDS: Record<string, TokenType> = {
  true: 'BOOL',
  false: 'BOOL',
  null: 'NULL',
  in: 'IN',
};

export class Lexer {
  private pos = 0;
  private input: string;

  constructor(input: string) {
    this.input = input;
  }

  tokenize(): Token[] {
    const tokens: Token[] = [];

    while (this.pos < this.input.length) {
      const char = this.input[this.pos];

      // Skip whitespace
      if (/\s/.test(char)) {
        this.pos++;
        continue;
      }

      // Numbers
      if (/\d/.test(char) || (char === '.' && /\d/.test(this.peek(1)))) {
        tokens.push(this.readNumber());
        continue;
      }

      // Strings
      if (char === '"' || char === "'") {
        tokens.push(this.readString(char));
        continue;
      }

      // Identifiers and keywords
      if (/[a-zA-Z_]/.test(char)) {
        tokens.push(this.readIdentifier());
        continue;
      }

      // Two-character operators
      const twoChar = this.input.slice(this.pos, this.pos + 2);
      if (twoChar === '==') {
        tokens.push({ type: 'EQ', value: '==', pos: this.pos });
        this.pos += 2;
        continue;
      }
      if (twoChar === '!=') {
        tokens.push({ type: 'NEQ', value: '!=', pos: this.pos });
        this.pos += 2;
        continue;
      }
      if (twoChar === '<=') {
        tokens.push({ type: 'LTE', value: '<=', pos: this.pos });
        this.pos += 2;
        continue;
      }
      if (twoChar === '>=') {
        tokens.push({ type: 'GTE', value: '>=', pos: this.pos });
        this.pos += 2;
        continue;
      }
      if (twoChar === '&&') {
        tokens.push({ type: 'AND', value: '&&', pos: this.pos });
        this.pos += 2;
        continue;
      }
      if (twoChar === '||') {
        tokens.push({ type: 'OR', value: '||', pos: this.pos });
        this.pos += 2;
        continue;
      }

      // Single-character operators
      const singleCharTokens: Record<string, TokenType> = {
        '.': 'DOT',
        ',': 'COMMA',
        '(': 'LPAREN',
        ')': 'RPAREN',
        '[': 'LBRACKET',
        ']': 'RBRACKET',
        '+': 'PLUS',
        '-': 'MINUS',
        '*': 'STAR',
        '/': 'SLASH',
        '%': 'PERCENT',
        '<': 'LT',
        '>': 'GT',
        '!': 'NOT',
        '?': 'QUESTION',
        ':': 'COLON',
      };

      if (singleCharTokens[char]) {
        tokens.push({ type: singleCharTokens[char], value: char, pos: this.pos });
        this.pos++;
        continue;
      }

      throw new Error(`Unexpected character '${char}' at position ${this.pos}`);
    }

    tokens.push({ type: 'EOF', value: null, pos: this.pos });
    return tokens;
  }

  private peek(offset: number): string {
    return this.input[this.pos + offset] ?? '';
  }

  private readNumber(): Token {
    const start = this.pos;
    let hasDecimal = false;

    while (this.pos < this.input.length) {
      const char = this.input[this.pos];
      if (/\d/.test(char)) {
        this.pos++;
      } else if (char === '.' && !hasDecimal) {
        hasDecimal = true;
        this.pos++;
      } else {
        break;
      }
    }

    const value = this.input.slice(start, this.pos);
    return {
      type: 'NUMBER',
      value,
      pos: start,
    };
  }

  private readString(quote: string): Token {
    const start = this.pos;
    this.pos++; // Skip opening quote

    let value = '';
    while (this.pos < this.input.length) {
      const char = this.input[this.pos];
      if (char === quote) {
        this.pos++; // Skip closing quote
        return { type: 'STRING', value, pos: start };
      }
      if (char === '\\') {
        this.pos++;
        const escaped = this.input[this.pos];
        const escapeMap: Record<string, string> = {
          n: '\n',
          t: '\t',
          r: '\r',
          '\\': '\\',
          '"': '"',
          "'": "'",
        };
        value += escapeMap[escaped] ?? escaped;
      } else {
        value += char;
      }
      this.pos++;
    }

    throw new Error(`Unterminated string starting at position ${start}`);
  }

  private readIdentifier(): Token {
    const start = this.pos;

    while (this.pos < this.input.length && /[a-zA-Z0-9_]/.test(this.input[this.pos])) {
      this.pos++;
    }

    const value = this.input.slice(start, this.pos);
    const keywordType = KEYWORDS[value];

    if (keywordType === 'BOOL') {
      return { type: 'BOOL', value: value === 'true', pos: start };
    }
    if (keywordType === 'NULL') {
      return { type: 'NULL', value: null, pos: start };
    }
    if (keywordType) {
      return { type: keywordType, value, pos: start };
    }

    return { type: 'IDENT', value, pos: start };
  }
}

File: rust/ais-rs/crates/ais-cel/src/lexer.rs
#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub kind: TokenKind,
    pub lexeme: String,
    pub pos: usize,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TokenKind {
    Identifier,
    Integer,
    Decimal,
    String,
    True,
    False,
    Null,
    In,
    Plus,
    Minus,
    Star,
    Slash,
    Percent,
    LParen,
    RParen,
    LBracket,
    RBracket,
    Comma,
    Dot,
    Question,
    Colon,
    Bang,
    EqEq,
    NotEq,
    Lt,
    Le,
    Gt,
    Ge,
    AndAnd,
    OrOr,
    Eof,
}

#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum LexError {
    #[error("unexpected character '{ch}' at {pos}")]
    UnexpectedCharacter { ch: char, pos: usize },
    #[error("unterminated string at {pos}")]
    UnterminatedString { pos: usize },
    #[error("invalid escape sequence at {pos}")]
    InvalidEscape { pos: usize },
}

pub fn tokenize(input: &str) -> Result<Vec<Token>, LexError> {
    let chars: Vec<char> = input.chars().collect();
    let mut index = 0usize;
    let mut tokens = Vec::new();

    while index < chars.len() {
        let ch = chars[index];
        if ch.is_whitespace() {
            index += 1;
            continue;
        }

        if ch.is_ascii_alphabetic() || ch == '_' {
            tokens.push(consume_identifier(&chars, &mut index));
            continue;
        }

        if ch.is_ascii_digit() {
            tokens.push(consume_number(&chars, &mut index));
            continue;
        }

        if ch == '\'' || ch == '"' {
            tokens.push(consume_string(&chars, &mut index)?);
            continue;
        }

        let token = tokenize_symbol(&chars, &mut index)?;
        tokens.push(token);
        index += 1;
    }

    tokens.push(Token {
        kind: TokenKind::Eof,
        lexeme: String::new(),
        pos: chars.len(),
    });

    Ok(tokens)
}

fn consume_identifier(chars: &[char], index: &mut usize) -> Token {
    let start = *index;
    *index += 1;
    while *index < chars.len() && (chars[*index].is_ascii_alphanumeric() || chars[*index] == '_') {
        *index += 1;
    }
    let lexeme: String = chars[start..*index].iter().collect();
    let kind = match lexeme.as_str() {
        "true" => TokenKind::True,
        "false" => TokenKind::False,
        "null" => TokenKind::Null,
        "in" => TokenKind::In,
        _ => TokenKind::Identifier,
    };
    Token {
        kind,
        lexeme,
        pos: start,
    }
}

fn consume_number(chars: &[char], index: &mut usize) -> Token {
    let start = *index;
    *index += 1;
    while *index < chars.len() && chars[*index].is_ascii_digit() {
        *index += 1;
    }

    let mut kind = TokenKind::Integer;
    if *index < chars.len() && chars[*index] == '.' {
        let dot = *index;
        *index += 1;
        if *index >= chars.len() || !chars[*index].is_ascii_digit() {
            *index = dot;
        } else {
            kind = TokenKind::Decimal;
            while *index < chars.len() && chars[*index].is_ascii_digit() {
                *index += 1;
            }
        }
    }

    let lexeme: String = chars[start..*index].iter().collect();
    Token {
        kind,
        lexeme,
        pos: start,
    }
}

fn consume_string(chars: &[char], index: &mut usize) -> Result<Token, LexError> {
    let quote = chars[*index];
    let start = *index;
    *index += 1;
    let mut out = String::new();
    let mut terminated = false;

    while *index < chars.len() {
        let current = chars[*index];
        if current == quote {
            *index += 1;
            terminated = true;
            break;
        }
        if current == '\\' {
            *index += 1;
            if *index >= chars.len() {
                return Err(LexError::UnterminatedString { pos: start });
            }
            let escaped = chars[*index];
            let decoded = decode_escape(escaped, *index)?;
            out.push(decoded);
            *index += 1;
            continue;
        }
        out.push(current);
        *index += 1;
    }

    if !terminated {
        return Err(LexError::UnterminatedString { pos: start });
    }

    Ok(Token {
        kind: TokenKind::String,
        lexeme: out,
        pos: start,
    })
}

fn decode_escape(escaped: char, pos: usize) -> Result<char, LexError> {
    let decoded = match escaped {
        'n' => '\n',
        'r' => '\r',
        't' => '\t',
        '\\' => '\\',
        '\'' => '\'',
        '"' => '"',
        _ => return Err(LexError::InvalidEscape { pos }),
    };
    Ok(decoded)
}

fn tokenize_symbol(chars: &[char], index: &mut usize) -> Result<Token, LexError> {
    let ch = chars[*index];
    let pos = *index;
    if let Some(token) = two_char_op(chars, index, ch, pos)? {
        return Ok(token);
    }
    one_char_op(ch, pos).ok_or(LexError::UnexpectedCharacter { ch, pos })
}

fn simple(kind: TokenKind, ch: char, pos: usize) -> Token {
    Token {
        kind,
        lexeme: ch.to_string(),
        pos,
    }
}

fn token_pair(kind: TokenKind, lexeme: &str, pos: usize) -> Token {
    Token {
        kind,
        lexeme: lexeme.to_string(),
        pos,
    }
}

fn two_char_op(chars: &[char], index: &mut usize, ch: char, pos: usize) -> Result<Option<Token>, LexError> {
    let token = match ch {
        '!' if matches_next(chars, pos, '=') => Some((TokenKind::NotEq, "!=")),
        '=' if matches_next(chars, pos, '=') => Some((TokenKind::EqEq, "==")),
        '<' if matches_next(chars, pos, '=') => Some((TokenKind::Le, "<=")),
        '>' if matches_next(chars, pos, '=') => Some((TokenKind::Ge, ">=")),
        '&' if matches_next(chars, pos, '&') => Some((TokenKind::AndAnd, "&&")),
        '|' if matches_next(chars, pos, '|') => Some((TokenKind::OrOr, "||")),
        '=' | '&' | '|' => return Err(LexError::UnexpectedCharacter { ch, pos }),
        _ => None,
    };
    if let Some((kind, lexeme)) = token {
        *index += 1;
        return Ok(Some(token_pair(kind, lexeme, pos)));
    }
    Ok(None)
}

fn one_char_op(ch: char, pos: usize) -> Option<Token> {
    let kind = match ch {
        '+' => TokenKind::Plus,
        '-' => TokenKind::Minus,
        '*' => TokenKind::Star,
        '/' => TokenKind::Slash,
        '%' => TokenKind::Percent,
        '(' => TokenKind::LParen,
        ')' => TokenKind::RParen,
        '[' => TokenKind::LBracket,
        ']' => TokenKind::RBracket,
        ',' => TokenKind::Comma,
        '.' => TokenKind::Dot,
        '?' => TokenKind::Question,
        ':' => TokenKind::Colon,
        '!' => TokenKind::Bang,
        '<' => TokenKind::Lt,
        '>' => TokenKind::Gt,
        _ => return None,
    };
    Some(simple(kind, ch, pos))
}

fn matches_next(chars: &[char], index: usize, expected: char) -> bool {
    chars.get(index + 1).copied() == Some(expected)
}

#[cfg(test)]
#[path = "lexer_test.rs"]
mod tests;
